### PR TITLE
Mime url prefix

### PR DESCRIPTION
**Pull request by regit**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/5174

Describe changes:
- Add scheme to URL now that Suricata supports multiple scheme
- Add some extensions to exe list
- Add link type in email

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 6462

File: src/output-json-email-common.c [hunk 1]
```diff
@@ -51,6 +51,7 @@
 
 #include "output-json.h"
 #include "output-json-email-common.h"
+#include <stdbool.h>
```
Comment (victorjulien) [inline src/output-json-email-common.c:54]: should already be in suricata-common.h

File: src/output-json-email-common.c [hunk 2]
```diff
@@ -295,17 +296,31 @@ static bool EveEmailLogJsonData(const Flow *f, void *state, void *vtx, uint64_t
...
+        bool has_ipv4_url = false;
+        bool has_exe_url = false;
         if (entity->url_list != NULL) {
             MimeDecUrl *url;
             for (url = entity->url_list; url != NULL; url = url->next) {
                 char *s = BytesToString((uint8_t *)url->url,
                                         (size_t)url->url_len);
                 if (s != NULL) {
                     jb_append_string(js_url, s);
+                    if (url->url_flags) {
```
Comment (victorjulien) [inline src/output-json-email-common.c:309]: formatting is off

Also, can it just be 3 lines like
`has_exe_url = ((url->url_flags & URL_IS_EXE) != 0);`

File: src/util-decode-mime.c [hunk 3]
```diff
@@ -1056,16 +1063,17 @@ static int FindUrlStrings(const uint8_t *line, uint32_t len,
                     SCLogDebug("Found url string");
 
                     /* First copy to temp URL string */
-                    tempUrl = SCMalloc(tokLen);
+                    tempUrl = SCMalloc(schemeStrLen + tokLen);
                     if (unlikely(tempUrl == NULL)) {
                         SCLogError(SC_ERR_MEM_ALLOC, "Memory allocation failed");
                         return MIME_DEC_ERR_MEM;
                     }
+                    memcpy(tempUrl, schemeStr, schemeStrLen);
```
Comment (victorjulien) [inline src/util-decode-mime.c:1071]: prefer switching to `SCCalloc` instead

**Review by victorjulien — CHANGES_REQUESTED** [review]

**Comment by suricata-qa** [comment]
Warning: no commits in this PR have specified the following ticket(s):
- 5174 - https://redmine.openinfosecfoundation.org/issues/5174

Please update the commit(s) and submit a new PR.

**Comment by victorjulien** [comment]
Shivani can you pick this one up and add some SV tests as well?

**Comment by inashivb** [comment]
> Shivani can you pick this one up and add some SV tests as well?

On it today.

File: src/util-decode-mime.c (hunk 3 above)
Comment (victorjulien) [reply src/util-decode-mime.c:1071]: looks like I misread this for a memset to 0

**Comment by inashivb** [comment]
Upon testing, this seems to be breaking the existing logging.
e.g. s-v test `smtp-extract-url-schemes` currently logs:
```
  "email": {
    "status": "PARSE_DONE",
    "from": "kKOEWPYxSvc@ajMvlQs1O4UA.com",
    "to": [
      "q7mp@Ku6FXfnLmo.com"
    ],
    "url": [
      "ftp://172.42.48.124",
      "https://facebook.com",
      "http://test-site.org/blah/123/",
      "http://google.com"
    ],
    "subject": "70meCK6P2y KvF 30cBys82YG OIrkmaK",
    "subject_md5": "0f99ded315de23e87d720f4c5c8fb7d4"
```

but with this patch, we see
```
  "email": {
    "status": "PARSE_DONE",
    "from": "kKOEWPYxSvc@ajMvlQs1O4UA.com",
    "to": [
      "q7mp@Ku6FXfnLmo.com"
    ],
    "has_ipv6_url": false,
    "has_ipv4_url": false,
    "has_exe_url": false,
    "url": [
      "ftp://ftp://172.42.48.124",
      "https://https://facebook.com",
      "http://http://test-site.org/blah/123/",
      "http://http://google.com"
    ],
    "subject": "70meCK6P2y KvF 30cBys82YG OIrkmaK",
    "subject_md5": "0f99ded315de23e87d720f4c5c8fb7d4"
  }
```
Notice the extra prefix in all urls..

**Comment by victorjulien** [comment]
Interesting, I think I've seen the issue somewhere as well, but can't remember where

**Comment by inashivb** [comment]
To me it seems like the url is getting copied w its prefix as prefix already seems to be a part of the `tok` https://github.com/OISF/suricata/blob/master/src/util-decode-mime.c#L1069
I cannot seem to reproduce the bug mentioned on the redmine ticket..
Is it for certain specific cases @regit ? Could you please share the pcap where this fails?

**Comment by regit** [comment]
272786908cf130b7d84324af141dbad6b7695a58 has been merged so the patch [500dcb1](https://github.com/OISF/suricata/pull/7103/commits/500dcb109381166824d99945645fc50430278149) is useless. I saw the MR when starting to work on it but wanted to have a simpler approach.

**Comment by regit** [comment]
It was global. I'm going to check with my pcap samples and suricata-verify ones.

**Comment by regit** [comment]
OK, got it: the patch introducing the scheme support breaks the logging if ever the configuration is not updated. Using master code and non updated configuration I got this output with the pcap from test from suricata-verify:
```
  "email": {
    "status": "PARSE_DONE",
    "from": "kKOEWPYxSvc@ajMvlQs1O4UA.com",
    "to": [
      "q7mp@Ku6FXfnLmo.com"
    ],
    "url": [
      "s://facebook.com",
      "://test-site.org/blah/123/",
      "://google.com"
    ]
  }
}
```

**Pull request closed without merging**